dune/grid/polyhedralgrid/grid.hh

```diff
@@ -0,0 +1,1207 @@
+        {
+          typedef std::map<int,int> vertexmap_t;
+          typedef typename vertexmap_t :: iterator iterator;
+
+          std::vector< vertexmap_t > cell_pts( dim*2 );
+
+          for (int hf=grid_.cell_facepos[ c ]; hf < grid_.cell_facepos[c+1]; ++hf)
+          {
+            const int f = grid_.cell_faces[ hf ];
+            const int faceTag = grid_.cell_facetag[ hf ];
+
+            for( int nodepos=grid_.face_nodepos[f]; nodepos<grid_.face_nodepos[f+1]; ++nodepos )
+            {
+              const int node = grid_.face_nodes[ nodepos ];
+              iterator it = cell_pts[ faceTag ].find( node );
+              if( it == cell_pts[ faceTag ].end() )
+              {
+                 cell_pts[ faceTag ].insert( std::make_pair( node, 1 ) );
+              }
+              else
+              {
+                // increase vertex reference counter
+                (*it).second++;
+              }
+            }
+          }
+
+          typedef std::map< int, std::set<int> > vertexlist_t;
+          vertexlist_t vertexList;
+
+          for( int faceTag = 0; faceTag<6; ++faceTag )
+          {
+            for( iterator it = cell_pts[ faceTag ].begin(),
+                 end = cell_pts[ faceTag ].end(); it != end; ++it )
+            {
+              // only consider vertices with one appearance
+              if( (*it).second == 1 )
+              {
+                vertexList[ (*it).first ].insert( faceTag );
+              }
+            }
+          }
+
+          cellVertices_[ c ].resize( vertexList.size() );
+          assert( int(vertexList.size()) == ( dim == 2 ) ? 4 : 8 );
+          for( auto it = vertexList.begin(), end = vertexList.end(); it != end; ++it )
+          {
+            assert( (*it).second.size() == dim );
+            KeyType key; key.fill( 4 ); // fill with 4 which is the first z coord
+
+            std::copy( (*it).second.begin(), (*it).second.end(), key.begin() );
+            auto vx = vertexFaceTags.find( key );
+            assert( vx != vertexFaceTags.end() );
+            // store node numbder on correct local position
+            cellVertices_[ c ][ (*vx).second ] = (*it).first ;
+          }
+        }
+      }
+      else
+      {
+        for (int c = 0; c < numCells; ++c)
+        {
+          std::set<int> cell_pts;
+          for (int hf=grid_.cell_facepos[ c ]; hf < grid_.cell_facepos[c+1]; ++hf)
+          {
+             int f = grid_.cell_faces[ hf ];
+             const int* fnbeg = grid_.face_nodes + grid_.face_nodepos[f];
+             const int* fnend = grid_.face_nodes + grid_.face_nodepos[f+1];
+             cell_pts.insert(fnbeg, fnend);
+          }
+
+          cellVertices_[ c ].resize( cell_pts.size() );
+          std::copy(cell_pts.begin(), cell_pts.end(), cellVertices_[ c ].begin() );
+        }
+      }
+
+      geomTypes_.resize(dim + 1);
+      GeometryType tmp;
+      for (int codim = 0; codim <= dim; ++codim) {
+        tmp.makeCube(dim - codim);
```

Comment:
This would be the same as CpGrid (except for also adding cubes for codims 1 and 2) and does not address the logical cartesian vs. general grid issue discussed above.